Add partner subsidiary id to upgrade URLs.

poisa opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Upgrade URLs will now globally include a subsidiary ID (if it exists). This value will come from a blog option called `jetpack_partner_subsidiary_id` that our hosting partners can set on the site themselves.

### Why?

Once we provision a Jetpack site for a hosting partner we customize the URLs we show users at the time of upgrading so they can follow the upgrade flow proposed by the partner that provisioned the site. 

Each of our partners gets a provisioning key that they use to provision sites and we use to keep track of stats and billing. This change allows a partner to configure upgrade URLs for many of their subsidiaries while using a single partner key (which allows us to keep stats and billing simple).

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This is a new feature which is part of the Partner Configuration project.
* Master Thread: parMIr-l2-p2
* The first partner that will use this feature is EuroDNS. 

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Checkout this PR.
* Run the following code snippet on your test site:

```php
<?php
function subsidiary_filter( $subsidiary ) {
	return 'from-filter-' . $subsidiary;
}

update_option( 'jetpack_partner_subsidiary_id', 'some-subsidiary-123', true );

add_filter( 'jetpack_partner_subsidiary_id', 'subsidiary_filter' );

```

> Tip: The easiest way I found to do this without running any CLI commands is to install the Hello Dolly plugin and use the plugin editor to paste the above code snippet right at the end of the `wp-content/plugins/hello-dolly/hello.php` file. Now, activate the plugin and the option will be set.

* Got to any page where Jetpack offers you an upgrade button. The plans page is a good choice.
* Inspect the upgrade link and you should see `&subsidiaryId=from-filter-some-subsidiary-123` appended at the end. The "from-filter-" string is added by the filter and the rest of the string is added by the option so if you see both, everything works.

In order to run the unit tests, a new test suite has been created:

```bash
yarn docker:phpunit --testsuite=partner-subsidiary
```

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add support for hosting partners to configure a subsidiary id to upgrade URLs.

poisa committed:
Add partner subsidiary id to upgrade URLs.

jeherve commented on _inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -266,6 +271,7 @@ function get_initial_state() {
 				'roles' => $stats_roles,
 			),
 			'aff' => Jetpack_Affiliate::init()->get_affiliate_code(),
+			'partnerSubsidiaryId'         => Jetpack_Partner_Subsidiary::init()->get_subsidiary_id(),
```
Spacing is quite tricky all over this array. Maybe we can take this opportunity to fix it (and thus not trip PHPCS when making changes)? Something like this:

```diff
diff --git a/_inc/lib/admin-pages/class.jetpack-react-page.php b/_inc/lib/admin-pages/class.jetpack-react-page.php
index d4708515e..5a52de181 100644
--- a/_inc/lib/admin-pages/class.jetpack-react-page.php
+++ b/_inc/lib/admin-pages/class.jetpack-react-page.php
@@ -236,32 +236,34 @@ class Jetpack_React_Page extends Jetpack_Admin_Page {
 		$current_user_data = jetpack_current_user_data();
 
 		return array(
-			'WP_API_root' => esc_url_raw( rest_url() ),
-			'WP_API_nonce' => wp_create_nonce( 'wp_rest' ),
-			'pluginBaseUrl' => plugins_url( '', JETPACK__PLUGIN_FILE ),
-			'connectionStatus' => array(
-				'isActive'  => Jetpack::is_active(),
-				'isStaging' => Jetpack::is_staging_site(),
-				'devMode'   => array(
+			'WP_API_root'                 => esc_url_raw( rest_url() ),
+			'WP_API_nonce'                => wp_create_nonce( 'wp_rest' ),
+			'pluginBaseUrl'               => plugins_url( '', JETPACK__PLUGIN_FILE ),
+			'connectionStatus'            => array(
+				'isActive'           => Jetpack::is_active(),
+				'isStaging'          => Jetpack::is_staging_site(),
+				'devMode'            => array(
 					'isActive' => ( new Status() )->is_development_mode(),
 					'constant' => defined( 'JETPACK_DEV_DEBUG' ) && JETPACK_DEV_DEBUG,
 					'url'      => site_url() && false === strpos( site_url(), '.' ),
 					'filter'   => apply_filters( 'jetpack_development_mode', false ),
 				),
-				'isPublic'	=> '1' == get_option( 'blog_public' ),
+				'isPublic'           => '1' == get_option( 'blog_public' ), // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
 				'isInIdentityCrisis' => Jetpack::validate_sync_error_idc_option(),
-				'sandboxDomain' => JETPACK__SANDBOX_DOMAIN,
+				'sandboxDomain'      => JETPACK__SANDBOX_DOMAIN,
 			),
-			'connectUrl' => $current_user_data['isConnected'] == false ? Jetpack::init()->build_connect_url( true, false, false ) : '',
-			'dismissedNotices' => $this->get_dismissed_jetpack_notices(),
-			'isDevVersion' => Jetpack::is_development_version(),
-			'currentVersion' => JETPACK__VERSION,
-			'is_gutenberg_available' => true,
-			'getModules' => $modules,
-			'rawUrl' => Jetpack::build_raw_urls( get_home_url() ),
-			'adminUrl' => esc_url( admin_url() ),
-			'stats' => array(
-				// data is populated asynchronously on page load
+			'connectUrl'                  => false == $current_user_data['isConnected'] // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
+				? Jetpack::init()->build_connect_url( true, false, false )
+				: '',
+			'dismissedNotices'            => $this->get_dismissed_jetpack_notices(),
+			'isDevVersion'                => Jetpack::is_development_version(),
+			'currentVersion'              => JETPACK__VERSION,
+			'is_gutenberg_available'      => true,
+			'getModules'                  => $modules,
+			'rawUrl'                      => Jetpack::build_raw_urls( get_home_url() ),
+			'adminUrl'                    => esc_url( admin_url() ),
+			'stats'                       => array(
+				// data is populated asynchronously on page load.
 				'data'  => array(
 					'general' => false,
 					'day'     => false,
@@ -270,18 +272,17 @@ class Jetpack_React_Page extends Jetpack_Admin_Page {
 				),
 				'roles' => $stats_roles,
 			),
-			'aff' => Jetpack_Affiliate::init()->get_affiliate_code(),
+			'aff'                         => Jetpack_Affiliate::init()->get_affiliate_code(),
 			'partnerSubsidiaryId'         => Jetpack_Partner_Subsidiary::init()->get_subsidiary_id(),
-			'settings' => $this->get_flattened_settings( $modules ),
-			'userData' => array(
-//				'othersLinked' => Jetpack::get_other_linked_admins(),
-				'currentUser'  => $current_user_data,
+			'settings'                    => $this->get_flattened_settings( $modules ),
+			'userData'                    => array(
+				'currentUser' => $current_user_data,
 			),
-			'siteData' => array(
-				'icon' => has_site_icon()
+			'siteData'                    => array(
+				'icon'                       => has_site_icon()
 					? apply_filters( 'jetpack_photon_url', get_site_icon_url(), array( 'w' => 64 ) )
 					: '',
-				'siteVisibleToSearchEngines' => '1' == get_option( 'blog_public' ),
+				'siteVisibleToSearchEngines' => '1' == get_option( 'blog_public' ), // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
 				/**
 				 * Whether promotions are visible or not.
 				 *
@@ -289,30 +290,30 @@ class Jetpack_React_Page extends Jetpack_Admin_Page {
 				 *
 				 * @param bool $are_promotions_active Status of promotions visibility. True by default.
 				 */
-				'showPromotions' => apply_filters( 'jetpack_show_promotions', true ),
-				'isAtomicSite' => jetpack_is_atomic_site(),
-				'plan' => Jetpack_Plan::get(),
-				'showBackups' => Jetpack::show_backups_ui(),
+				'showPromotions'             => apply_filters( 'jetpack_show_promotions', true ),
+				'isAtomicSite'               => jetpack_is_atomic_site(),
+				'plan'                       => Jetpack_Plan::get(),
+				'showBackups'                => Jetpack::show_backups_ui(),
 			),
-			'themeData' => array(
+			'themeData'                   => array(
 				'name'      => $current_theme->get( 'Name' ),
 				'hasUpdate' => (bool) get_theme_update_available( $current_theme ),
 				'support'   => array(
-					'infinite-scroll' => current_theme_supports( 'infinite-scroll' ) || in_array( $current_theme->get_stylesheet(), $inf_scr_support_themes ),
+					'infinite-scroll' => current_theme_supports( 'infinite-scroll' ) || in_array( $current_theme->get_stylesheet(), $inf_scr_support_themes, true ),
 				),
 			),
-			'locale' => Jetpack::get_i18n_data_json(),
-			'localeSlug' => join( '-', explode( '_', get_user_locale() ) ),
-			'jetpackStateNotices' => array(
-				'messageCode' => Jetpack::state( 'message' ),
-				'errorCode' => Jetpack::state( 'error' ),
+			'locale'                      => Jetpack::get_i18n_data_json(),
+			'localeSlug'                  => join( '-', explode( '_', get_user_locale() ) ),
+			'jetpackStateNotices'         => array(
+				'messageCode'      => Jetpack::state( 'message' ),
+				'errorCode'        => Jetpack::state( 'error' ),
 				'errorDescription' => Jetpack::state( 'error_description' ),
 			),
-			'tracksUserData' => Jetpack_Tracks_Client::get_connected_user_tracks_identity(),
-			'currentIp' => function_exists( 'jetpack_protect_get_ip' ) ? jetpack_protect_get_ip() : false,
-			'lastPostUrl' => esc_url( $last_post ),
+			'tracksUserData'              => Jetpack_Tracks_Client::get_connected_user_tracks_identity(),
+			'currentIp'                   => function_exists( 'jetpack_protect_get_ip' ) ? jetpack_protect_get_ip() : false,
+			'lastPostUrl'                 => esc_url( $last_post ),
 			'externalServicesConnectUrls' => $this->get_external_services_connect_urls(),
-			'calypsoEnv' => Jetpack::get_calypso_env(),
+			'calypsoEnv'                  => Jetpack::get_calypso_env(),
 		);
 	}
 
@@ -347,7 +348,7 @@ class Jetpack_React_Page extends Jetpack_Admin_Page {
  * @return array
  */
 function jetpack_current_user_data() {
-	$current_user = wp_get_current_user();
+	$current_user  = wp_get_current_user();
 	$is_master_user = $current_user->ID == Jetpack_Options::get_option( 'master_user' );
 	$dotcom_data    = Jetpack::get_connected_user_data();
 	// Add connected user gravatar to the returned dotcom_data.

```

jeherve reviewed (changes requested):
This looks good to me and should be good to merge. My only comments are about coding standards.

Could you add your new file to the PHPCS whitelist here? This will also give you the opportunity to make a few changes to the file to fix any PHPCS warnings that may pop up. Something like this:

```diff
diff --git a/bin/phpcs-whitelist.js b/bin/phpcs-whitelist.js
index 03cd49286..ca48d3f6b 100644
--- a/bin/phpcs-whitelist.js
+++ b/bin/phpcs-whitelist.js
@@ -4,6 +4,7 @@ module.exports = [
 	'3rd-party/class.jetpack-amp-support.php',
 	'3rd-party/woocommerce-services.php',
 	'class.jetpack-gutenberg.php',
+	'class.jetpack-partner-subsidiary.php',
 	'class.jetpack-plan.php',
 	'extensions/',
 	'functions.global.php',
diff --git a/class.jetpack-partner-subsidiary.php b/class.jetpack-partner-subsidiary.php
index e328f15f8..bdb68e2c1 100644
--- a/class.jetpack-partner-subsidiary.php
+++ b/class.jetpack-partner-subsidiary.php
@@ -1,27 +1,36 @@
-<?php
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
+/**
+ * Utilities to handle Partner Subsidiary IDs.
+ * Those IDs allow partners to customize the Jetpack onboarding experience.
+ *
+ * @since 8.1.0
+ *
+ * @package Jetpack
+ */
+
+use Automattic\Jetpack\Status;
 
 if ( ! defined( 'ABSPATH' ) ) {
 	// Exit if accessed directly.
 	exit;
 }
 
-use Automattic\Jetpack\Status;
-
 /**
  * This class introduces routines to get a partner subsidiary id, that might be obtained from:
  * - a `jetpack_partner_subsidiary_id` option in the WP database
  * - a subsidiary id returned by a filter bound to the `jetpack_partner_subsidiary_id` filter hook
  *
- * @since 8.0.0
+ * @since 8.1.0
  */
 class Jetpack_Partner_Subsidiary {
 
 	/**
 	 * Singleton instance.
 	 *
-	 * @since 8.0.0
+	 * @since 8.1.0
+	 *
 	 * @var Jetpack_Partner_Subsidiary This class instance.
-	 **/
+	 */
 	private static $instance = null;
 
 	/**
@@ -36,7 +45,7 @@ class Jetpack_Partner_Subsidiary {
 	/**
 	 * Initializes the class or returns the singleton
 	 *
-	 * @since 8.0.0
+	 * @since 8.1.0
 	 *
 	 * @return Jetpack_Partner_Subsidiary | false
 	 */
@@ -50,7 +59,7 @@ class Jetpack_Partner_Subsidiary {
 	/**
 	 * Returns the partner subsidiary id from database after filtering it.
 	 *
-	 * @since 8.0.0
+	 * @since 8.1.0
 	 *
 	 * @return string The affiliate code.
 	 */
@@ -68,7 +77,7 @@ class Jetpack_Partner_Subsidiary {
 	/**
 	 * Returns the passed URL with the partner subsidiary id code added as a URL query arg.
 	 *
-	 * @since @since 8.0.0
+	 * @since @since 8.1.0
 	 *
 	 * @param string $url The URL where the partner subsidiary id will be added.
 	 *

```

enejb commented on class.jetpack.php:
```diff
@@ -4622,6 +4622,12 @@ function build_connect_url( $raw = false, $redirect = false, $from = false, $reg
 		$url = Jetpack_Affiliate::init()->add_code_as_query_arg( $url );
 
+		// Ensure that class to get the partner subsidiary id is loaded.
+		if ( ! class_exists( 'Jetpack_Partner_Subsidiary' ) ) {
```
Instead of doing this. Have we though about implement it as a filter instead?

poisa replied:
It's implemented as a filter (and an option) here:

https://github.com/Automattic/jetpack/blob/63d5b6f979e6699c757581149bc3dd7305244c08/class.jetpack-partner-subsidiary.php#L66-L74

enejb replied:
What I was trying to get to is instead of doing 
```
I think instead of doing 

$url = Jetpack_Affiliate::init()->add_code_as_query_arg( $url );
$url = Partner::init()->add_subsidiary_id_as_query_arg( $url );
return $raw ? esc_url_raw( $url ) : esc_url( $url );
```

I think we should do something like the following. 
```
$url = $raw ? esc_url_raw( $url ) : esc_url( $url )
return apply_filters( 'jetpack_build_connection_url', $url, $raw );
```
and then 
on add the filters after all the plugin have loaded. 
```
add_filter( 'jetpack_build_connection_url', array( 'Jetpack_Affiliate', 'add_code_as_query_arg' );
add_filter( 'jetpack_build_connection_url', array( 'Partner', 'add_subsidiary_id_as_query_arg' );
```

enejb commented on class.jetpack-partner-subsidiary.php:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @since 8.0.0
+ */
+class Jetpack_Partner_Subsidiary {
```
Instead of adding this new class in Jetpack can we add it as a package? 
So that it could potentially be used by other plugins down the line?

poisa replied:
This makes sense. Currently, Jetpack Start only provisions Jetpack sites but we also side-load Akismet and VaultPress. We don't know if we are ever going to provision anything other than Jetpack but if we do it makes sense to put this in a place that can be used.

jeherve reviewed (changes requested):
Thinking about it some more, and since the class does exactly the same thing as the existing Affiliate class, I wonder if we shouldn't merge those 2 into a single class that could be used for both affliliates and partner IDs. That may avoid duplicating code unnecessarily.

poisa commented:
>I wonder if we shouldn't merge those 2 into a single class that could be used for both affliliates and partner IDs.

Just for context, the Subsidiary id is strictly a Jetpack Start concept and it "lives" in the `jetpack-start` lib in the WPCOM repo. The Affiliate code lives in the `jetpack-redirects` lib in the WPCOM repo. So strictly speaking they are different things. Based on @enejb's comment I'm leaning toward making this into a package. as it will be easier to use if we start to provision plugins other than Jetpack.

poisa commented:
Hi folks. I've updated the PR to use a package. I didn't find any docs on how to create one so I just used an existing one as an example. The trouble is that the autoloader is not recognizing the package so I'm not being able to neither load the class nor run the tests. I've posted more details here: p1574961969189500-slack-CDLH4C1UZ

Any help will be greatly appreciated!

jeherve commented:
> The trouble is that the autoloader is not recognizing the package so I'm not being able to neither load the class nor run the tests. 

I think you may have to add the new package to `composer.json`, since we need it in Jetpack:
https://github.com/Automattic/jetpack/blob/93fe3c93fdb87c95b2f8785f280435c6b315c0ad/composer.json#L10

> So strictly speaking they are different things. 

Right. They do the exact same thing, though. The only difference is the option / filter name. What do you think about having a single package like the one you've created, that would be able to handle both options / filters? 
Instead of having `Partner::init()->get_subsidiary_id()` and `Jetpack_Affiliate::init()->get_affiliate_code()`, we would have `( new Partner() )->get_partner_code( 'subsidiary' )` and `( new Partner() )->get_partner_code( 'affiliate' )`, for example?

poisa commented:
@jeherve @enejb I've updated the PR with the requested changes. I've attempted some testing but given that Jetpack DNA package tests have no WordPress environment setup, even the simplest tests require mocking a fair bit of WP functionality. This is no minor task and I still need to find a proper way to do this (mocking every single WP function that's called will result in very brittle tests).

Because of this, I'd rather have you comment on the core functionality of the package first, and I'll tackle testing once you give me the green light that the core is all good.

poisa commented:
I updated the PR with unit tests using the method described by @gravityrail here: p1575655532125700-slack-CQN0YDYM7

He requested that I attempt to automate creating the database instance using Docker so that it doesn't require one in the dev's machine. I'm going to attempt this on Monday. But as of the latest commit the package is functional and has tests.

jeherve commented:
It looks like this is now going to need a rebase.

poisa commented:
> It looks like this is now going to need a rebase.

Done!

poisa commented:
OK, I've rebased, squashed, and added the functionality @gravityrail requested.

Unit tests can now be run using the container's MySQL database by doing this:

```bash
yarn docker:compose exec wordpress bash -c 'cd /var/www/html/wp-content/plugins/jetpack/packages/partner && phpunit'
```

Please let me know what's next so that we can get this out the door. I'm going AFK on tuesday 17th but need to finish this before as we promised EuroDNS this feature.

jeherve commented:
Looks like the tests are failing right now. Here is an example:
https://travis-ci.org/Automattic/jetpack/jobs/622962639?utm_medium=notification&utm_source=github_status

poisa commented:
There are two reasons tests are failing: 

* I hadn't added `composer.lock` in the commit because it was referencing my this branch's name as the package version and this seems wrong. I've included this file and now those tests pass.

https://github.com/Automattic/jetpack/blob/0939b8b68cc02632c4318516d085dafcc86836a9/composer.lock#L351-L352

* Travis is attempting to run my unit tests using the docker database config and it is not being able to connect.

There are no docs for authoring Jetpack DNA packages and this is the only package that uses a docker MySQL container so at this point I'm basically guessing what to do. Dan requests that this package be updated to use it's own docker db (as opposed to the Jetpack Docker db which is currently being used). Unless I find a way to make Travis CI understand that this package needs tests to be ran differently they will never pass.

I'm going to need a bit more guidance in order to complete this patch successfully:

1) Can you confirm that the approach taken in the package works? If you can confirm this, I can go ahead and add some documentation to the README.md. I'd rather document the package once I know for sure that the code is up to par.

2) Is what Dan requests possible with Travis CI or will I be running around in circles? If it is indeed possible, can you provide some pointers on how to achieve it? I _think_ I can figure out the Docker side of things but I'm unfamiliar with Travis.

jeherve commented:
> There are no docs for authoring Jetpack DNA packages and this is the only package that uses a docker MySQL container so at this point I'm basically guessing what to do.

@tyxla @zinigor Do you have any ideas on how to approach this? That's a first here, so we should find a solution that we can follow for other packages in the future.

kraftbj commented:
It looks like the DB is only being used as part of the `test_partner_codes_are_added_to_connect_url` test.

Looking at it, that is really testing out core Jetpack functionality (the build URL function) and may better live in the core Jetpack testing suite. The unit tests for the package, IMO, should be testing the actual functionality of the package while the consumer (Jetpack) should be testing to confirm what they're expecting is still happening.

That should let us simplify things by not needing a separate database or am I missing something?

kraftbj committed:
Update tests to not require db

tyxla reviewed (commented):
I took a quick look too, and second what @kraftbj already suggested. Is there any good reason to run these tests in a separate docker instance?

kraftbj committed:
Remove deleted file

kraftbj commented:
A proposed alternative way of testing #14208 (that would merge into this branch).

Also, this PR is throwing a fatal error in the admin. I've fixed in 14208 to get clean tests.
```
[11-Dec-2019 16:13:36 UTC] PHP Fatal error:  require_once(): Failed opening required '/home/travis/build/Automattic/jetpack/class.jetpack-affiliate.php' (include_path='.:/home/travis/.phpenv/versions/7.0.25/share/pear') in /home/travis/build/Automattic/jetpack/load-jetpack.php on line 69
[11-Dec-2019 16:13:36 UTC] PHP Stack trace:
[11-Dec-2019 16:13:36 UTC] PHP   1. {main}() /home/travis/wordpress/wp-admin/index.php:0
[11-Dec-2019 16:13:36 UTC] PHP   2. require_once() /home/travis/wordpress/wp-admin/index.php:10
[11-Dec-2019 16:13:36 UTC] PHP   3. require_once() /home/travis/wordpress/wp-admin/admin.php:34
[11-Dec-2019 16:13:36 UTC] PHP   4. require_once() /home/travis/wordpress/wp-load.php:37
[11-Dec-2019 16:13:36 UTC] PHP   5. require_once() /home/travis/wordpress/wp-config.php:90
[11-Dec-2019 16:13:36 UTC] PHP   6. include_once() /home/travis/wordpress/wp-settings.php:360
[11-Dec-2019 16:13:36 UTC] PHP   7. require_once() /home/travis/build/Automattic/jetpack/jetpack.php:146```

poisa commented:
>That should let us simplify things by not needing a separate database or am I missing something?

It definitely simplifies things and you're effort is greatly appreciated! I've gone ahead and merged this branch with yours.

>Is there any good reason to run these tests in a separate docker instance?

This was Dan's request. He was aiming to have some sort of test skeleton for Jetpack DNA packages where you could test them independently of the Jetpack plugin. Since the merged changes make it so that this doesn't require any database I think this is not applicable to this package.

Just got a ping from EuroDNS today (600&cid=CJZ8FMX8E-slack-CJZ8FMX8E) asking for the status on this. Is there an ETA on when this will be shipped?

jeherve commented:
> Just got a ping from EuroDNS today (600&cid=CJZ8FMX8E-slack-CJZ8FMX8E) asking for the status on this. Is there an ETA on when this will be shipped?

The next Jetpack release is scheduled for January 14, 2020. I think this can make it in.

gravityrail commented:
@poisa I am ok with merging @kraftbj's patch that avoids a dependency on MySQL. I had a good discussion with him about it this morning, and here's where I stand:

Ultimately what we want is good tests, good code and lots of coverage.

If you find yourself mocking a lot of WP functionality, like add_action etc. then probably you have a brittle test and you do actually need to test against a "real" WP instance with a real database. 

If you can get away with light stubbing/mocking then you probably don't need a WP instance for your test.

In reality, most packages are going to need a bit of both. At least one "integration" style test that ensures that everything comes together nicely, and unit tests that make assertions on message passing between objects.

It is up to the developer to exercise good judgement about whether their tests are going to cause more trouble than they're worth. The one thing I really don't want in this case is an ideological commitment one way or the other, mock vs. integrate.

gravityrail commented:
Like, if your package does something that modifies WP's front-end output, and you don't have at least one test that checks that it works with a "real" WP page render, then you probably don't have enough coverage.

poisa committed:
Add README.md

jeherve commented on class.jetpack.php:
```diff
@@ -4615,14 +4617,8 @@ function build_connect_url( $raw = false, $redirect = false, $from = false, $reg
-
-		return $raw ? esc_url_raw( $url ) : esc_url( $url );
+		$url = $raw ? esc_url_raw( $url ) : esc_url( $url );
+		return apply_filters( 'jetpack_build_connection_url', $url, $raw );
```
Could you add a docblock here, to make our parser happy and allow us to automatically generate documentation on developer.jetpack.com? Something like this:

```suggestion
		/**
		 * Filter the URL used when connecting a user to a WordPress.com account.
		 *
		 * @since 8.1.0
		 *
		 * @param string $url Connection URL.
		 * @param bool   $raw If true, URL will not be escaped.
		 */
		return apply_filters( 'jetpack_build_connection_url', $url, $raw );
```

jeherve commented on modules/plugin-search.php:
```diff
@@ -1,6 +1,7 @@
 <?php
 
 use Automattic\Jetpack\Constants;
+use Automattic\Jetpack\Partner;
```
Good news! It looks like we can remove the whole `get_upgrade_url` function since 5308f676f40d22b44c5999e3f5fc72bfe8eb1524, it isn't used anywhere anymore. It's my bad, I should have removed it back then, when removing the only use of it. 

You can safely delete it, thus making this file a bit simpler.

jeherve commented on packages/partner/.gitignore:
```diff
@@ -0,0 +1 @@
+wordpress
```
Maybe we can get rid of this now?

jeherve commented on composer.lock:
```diff
@@ -4,16 +4,15 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
```
I think we may have to composer update again to remove the references to the testing tools we don't need anymore.

jeherve commented on packages/partner/src/class-partner.php:
```diff
@@ -0,0 +1,154 @@
+	 * @param string $url The URL where the partner subsidiary id will be added.
+	 *
+	 * @return string The passed URL with the partner code added.
+	 * @since @since 8.1.0
```
```suggestion
	 * @since 8.1.0
```

jeherve commented on packages/partner/src/class-partner.php:
```diff
@@ -0,0 +1,154 @@
+					'jetpack_partner_subsidiary_id',
+					get_option( 'jetpack_partner_subsidiary_id', '' )
+				);
+				/**
```
Looks like the docblock stayed after the filter; could you move it before the return so as to not trip up the parser?

jeherve commented on packages/partner/tests/php/test_Partner.php:
```diff
@@ -0,0 +1,87 @@
+	 *
+	 * @throws Monkey\Expectation\Exception\ExpectationArgsRequired
+	 */
+	public function test_partner_code_is_empty_by_defalt( $code_type, $option_name, $query_string_name ) {
```
Nitpicking here:

```suggestion
	public function test_partner_code_is_empty_by_default( $code_type, $option_name, $query_string_name ) {
```

jeherve commented on tests/php/test_class.jetpack-affiliate.php:
```diff
@@ -1,47 +0,0 @@
-<?php
```
Since we're removing the file, you may have to remove it from `phpunit.xml.dist` too.
https://github.com/Automattic/jetpack/blob/01b62d9b421c3fa7ae97c00f434a4c3343d2d4af/phpunit.xml.dist#L82

poisa committed:
[not verified] Add header to stub file class.jetpack-affiliate.php

poisa committed:
Add dockblock to jetpack_build_connection_url filter

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

poisa committed:
Merge branch 'add/upgrade-url-subsidiary-id' of github.com:Automattic/jetpack into add/upgrade-url-subsidiary-id

poisa committed:
Remove get_upgrade_url function

poisa committed:
Removed wordpress from .gitignore as not WordPress environment is used anymore

poisa committed:
Update composer.lock to remove unused dependencies

poisa committed:
Moved filter docblock to the correct place

poisa committed:
Remove affiliate test suite from phpunit.xml.dist

poisa committed:
Remove duplicate @since annotation

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

poisa committed:
Fix typo in test function test_partner_code_is_empty_by_default

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

jeherve committed:
Remove unneeded gitignore file

jeherve committed:
Remove unneeded break

There is no need to break after a return.

jeherve reviewed (approved):
This looks good to me now. It should be good to merge now!
